docs/hugo/content/design/ADR-2023-07-Goal-Seeking-KeyVaults/_index.md

```diff
@@ -0,0 +1,154 @@
+
+### Option 3: Sidestep the Problem
+
+If we attempt to create a KeyVault and encounter a name collision, we could automatically select a different AzureName for the KeyVault and publish that as a ConfigMap. The application could then be configured to use the new name.
+
+* CON: May be surprising to users
+* CON: If the KeyVault was deleted by a malicious actor, we'd respond by starting afresh with a blank KeyVault, resulting in a longer outage than necessary.
+* CON: Only applications running in the cluster would be able to use the new KeyVault name. Applications running outside the cluster would continue to use the old name.
+* CON: Requires that applications running in the cluster load their KeyVault name from a configmap, rather than hard coding it via other config.
+
+
+### Option 4: Automatic Recovery
+
+Remove the `createMode` property entirely and automatically recover any soft-deleted KeyVaults that we encounter.
+
+This is a natural extension of ASO as a maintainer of the declared goal state. If the goal state is to have a KeyVault, and there is a soft-deleted KeyVault, then a natural action is to recover the soft-deleted KeyVault.
+
+* PRO: We can remove the property using existing generator config (though it would be a breaking change.)
+* PRO: We can implement the behaviour using existing extension points.
+* PRO: Likely to be the behaviour wanted by most users.
+* CON: No way to opt-out
+
+### Option 5: New Property
+
+Remove the existing `createMode` property and create a new property to configure the behaviour, say `reconciliationMode`:
+
+`block` - Create a new KeyVault if there is no existing KeyVault with the same name. If there is an existing KeyVault with the same name, block creation.
+
+`recover` - Create a new KeyVault if there is no existing KeyVault with the same name. If there is an existing KeyVault with the same name, recover it. 
+
+We could specify `recover` as the default.
+
+* PRO: We're not redefining `createMode` so less possiblity for confusion there.
+* PRO: A new property can have exactly the semantics we want.
+* CON: Users may be surprised that `createMode` is not supported.
+* CON: We don't have a general purpose technique for introducing new custom properties and would need to add one.
+
+
+### Option 6: New Property with Purge
+
+As for **Option 6**, but we include an option to purge the KeyVault if it exists.
+
+* PRO: Allows users to configure all possible behaviours.
+* CON: Dangerous, as we could accidentally purge a KeyVault that we weren't intended to purge.
+* CON: The KeyVault Program Group doesn't provide this, likely for good reason.
+
+If someone soft-deleted a KeyVault (whether maliciously or accidentally), automatically purging that KeyVault would make any existing secrets unrecoverable, compounding any problems. To suggest *this would be a bad thing* would be somewhat of an understatement.
+
+
+## Decision
+
+Reccommendation: Option 1 - Reinterpretation of createMode
```

Comment:
I think I generally agree with this conclusion although I think that we should consider doing what I suggested on option 1 (adding options) rather than changing the meaning of existing options.